### PR TITLE
Share unified UTF-8/16 parsing with BigInteger

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Follow up of #85978 . Should really be in same PR.

Copied latest CoreLib implementation and polyfill internal interfaces.
This allows sharing UTF unified parsing and formatting code.

Originally I tried this with heavy ifdefs and tricks like `using TChar = System.Char;`. Then I realized that all the internal members can be polyfilled in S.R.Numerics, with very little overhead for UTF-16. This also unblocks UTF-8 adoption for BigInteger.

Wanna check with @stephentoub for whether this design is desired.

### CODE REVIEW

File: src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    // Polyfill CoreLib internal interfaces and methods
+    // Define necessary members only
+
+    internal interface IUtfChar<TSelf>
```
Comment (huoyaoyuan) [inline src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs:14]: CoreLib comments indicate this interface to be a workaround for inliner limitation. Is it resolved now? Then we can move to `IBinaryInteger` and remove the polyfill.

Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs:14]: It looks like resolved with simple benchmarking on Int*.Parse:

| Method       | Job        | Toolchain                  | value                | Mean      | Error     | StdDev    | Median    | Min       | Max       | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------- |----------- |--------------------------- |--------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|----------:|------------:|
| Parse        | Job-MGZKSL | \BinaryInteger\corerun.exe | -9223372036854775808 | 10.932 ns | 0.1064 ns | 0.0996 ns | 10.935 ns | 10.747 ns | 11.130 ns |  1.02 |    0.02 |         - |          NA |
| Parse        | Job-TCKMSW | \base\corerun.exe          | -9223372036854775808 | 10.725 ns | 0.2329 ns | 0.2178 ns | 10.693 ns | 10.366 ns | 11.095 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParse     | Job-MGZKSL | \BinaryInteger\corerun.exe | -9223372036854775808 | 11.103 ns | 0.1603 ns | 0.1499 ns | 11.035 ns | 10.923 ns | 11.374 ns |  1.02 |    0.03 |         - |          NA |
| TryParse     | Job-TCKMSW | \base\corerun.exe          | -9223372036854775808 | 10.898 ns | 0.2303 ns | 0.2464 ns | 10.810 ns | 10.653 ns | 11.524 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| ParseSpan    | Job-MGZKSL | \BinaryInteger\corerun.exe | -9223372036854775808 | 11.050 ns | 0.0883 ns | 0.0826 ns | 11.030 ns | 10.945 ns | 11.208 ns |  1.00 |    0.01 |         - |          NA |
| ParseSpan    | Job-TCKMSW | \base\corerun.exe          | -9223372036854775808 | 11.100 ns | 0.1081 ns | 0.1011 ns | 11.091 ns | 10.927 ns | 11.268 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParseSpan | Job-MGZKSL | \BinaryInteger\corerun.exe | -9223372036854775808 | 10.851 ns | 0.1740 ns | 0.1453 ns | 10.859 ns | 10.575 ns | 11.110 ns |  1.01 |    0.01 |         - |          NA |
| TryParseSpan | Job-TCKMSW | \base\corerun.exe          | -9223372036854775808 | 10.698 ns | 0.0577 ns | 0.0512 ns | 10.692 ns | 10.621 ns | 10.786 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| Parse        | Job-MGZKSL | \BinaryInteger\corerun.exe | 12345                |  4.582 ns | 0.0285 ns | 0.0253 ns |  4.589 ns |  4.535 ns |  4.624 ns |  1.01 |    0.01 |         - |          NA |
| Parse        | Job-TCKMSW | \base\corerun.exe          | 12345                |  4.558 ns | 0.0157 ns | 0.0139 ns |  4.561 ns |  4.526 ns |  4.574 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParse     | Job-MGZKSL | \BinaryInteger\corerun.exe | 12345                |  4.643 ns | 0.0555 ns | 0.0520 ns |  4.627 ns |  4.588 ns |  4.772 ns |  1.09 |    0.02 |         - |          NA |
| TryParse     | Job-TCKMSW | \base\corerun.exe          | 12345                |  4.265 ns | 0.0750 ns | 0.0701 ns |  4.232 ns |  4.184 ns |  4.407 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| ParseSpan    | Job-MGZKSL | \BinaryInteger\corerun.exe | 12345                |  4.233 ns | 0.0850 ns | 0.0795 ns |  4.206 ns |  4.137 ns |  4.371 ns |  0.95 |    0.02 |         - |          NA |
| ParseSpan    | Job-TCKMSW | \base\corerun.exe          | 12345                |  4.474 ns | 0.0274 ns | 0.0256 ns |  4.478 ns |  4.432 ns |  4.520 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParseSpan | Job-MGZKSL | \BinaryInteger\corerun.exe | 12345                |  4.646 ns | 0.0341 ns | 0.0302 ns |  4.650 ns |  4.589 ns |  4.684 ns |  0.99 |    0.01 |         - |          NA |
| TryParseSpan | Job-TCKMSW | \base\corerun.exe          | 12345                |  4.679 ns | 0.0773 ns | 0.0723 ns |  4.646 ns |  4.611 ns |  4.800 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| Parse        | Job-MGZKSL | \BinaryInteger\corerun.exe | 9223372036854775807  | 10.146 ns | 0.0732 ns | 0.0685 ns | 10.185 ns |  9.996 ns | 10.207 ns |  1.01 |    0.01 |         - |          NA |
| Parse        | Job-TCKMSW | \base\corerun.exe          | 9223372036854775807  | 10.081 ns | 0.0829 ns | 0.0776 ns | 10.066 ns |  9.995 ns | 10.218 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParse     | Job-MGZKSL | \BinaryInteger\corerun.exe | 9223372036854775807  |  9.959 ns | 0.0401 ns | 0.0335 ns |  9.971 ns |  9.865 ns |  9.997 ns |  0.99 |    0.01 |         - |          NA |
| TryParse     | Job-TCKMSW | \base\corerun.exe          | 9223372036854775807  | 10.099 ns | 0.0562 ns | 0.0499 ns | 10.096 ns | 10.044 ns | 10.204 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| ParseSpan    | Job-MGZKSL | \BinaryInteger\corerun.exe | 9223372036854775807  | 10.030 ns | 0.0773 ns | 0.0686 ns | 10.043 ns |  9.939 ns | 10.182 ns |  0.99 |    0.01 |         - |          NA |
| ParseSpan    | Job-TCKMSW | \base\corerun.exe          | 9223372036854775807  | 10.129 ns | 0.1174 ns | 0.0916 ns | 10.089 ns | 10.025 ns | 10.259 ns |  1.00 |    0.00 |         - |          NA |
|              |            |                            |                      |           |           |           |           |           |           |       |         |           |             |
| TryParseSpan | Job-MGZKSL | \BinaryInteger\corerun.exe | 9223372036854775807  | 10.054 ns | 0.0628 ns | 0.0588 ns | 10.059 ns |  9.957 ns | 10.166 ns |  0.96 |    0.02 |         - |          NA |
| TryParseSpan | Job-TCKMSW | \base\corerun.exe          | 9223372036854775807  | 10.503 ns | 0.2312 ns | 0.2270 ns | 10.460 ns | 10.150 ns | 11.025 ns |  1.00 |    0.00 |         - |          NA |

File: src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        : IEquatable<TSelf>
```
Comment (huoyaoyuan) [inline src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs:15]:
```suggestion
    internal interface IUtfChar<TSelf> :
        IEquatable<TSelf>
```

**Comment by huoyaoyuan** [comment]
Test failures look unrelated. No integer/biginteger related.

File: src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs (hunk 1 above)
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs]: Hmm `IBinaryInteger` may be unfriendly for mono.

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Runtime.Numerics/src/System/Number.Polyfill.cs]: Yeah the issue was with Mono and its something on their backlog to get resolved.

**Comment by tannergooding** [comment]
CC. @vargaz, another LLVM OOM. It's not clear to me which line indicates the failing assembly so I can know what needs to be skipped until the final PR can get merged?
```
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : Failed to compile C:\helix\work\workitem\e\wasm_build\obj\wasm\for-build\aot-instances.dll.bc -> C:\helix\work\workitem\e\wasm_build\obj\wasm\for-build\aot-instances.dll.o [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : C:\helix\work\workitem\e\publish>C:\Windows\System32\chcp.com 65001 1>nul  [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : C:\helix\work\workitem\e\publish>setlocal [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : C:\helix\work\workitem\e\publish>set errorlevel=dummy  [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : C:\helix\work\workitem\e\publish>set errorlevel=  [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : C:\helix\work\workitem\e\publish>"C:\helix\work\correlation\build\\wasi-sdk\\bin\clang.exe" "@C:\helix\work\correlation\build\microsoft.netcore.app.runtime.wasi-wasm\runtimes\wasi-wasm\native\src\wasi-default.rsp" "@C:\helix\work\workitem\e\wasm_build\obj\wasm\for-build\wasi-compile-bc.rsp" -c -o "C:\Users\ContainerAdministrator\AppData\Local\Temp\tmpa2qi3t.tmp" "C:\helix\work\workitem\e\wasm_build\obj\wasm\for-build\aot-instances.dll.bc"  [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : clang version 16.0.0 [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : Target: wasm32-unknown-wasi [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : Thread model: posix [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : InstalledDir: C:/helix/work/correlation/build//wasi-sdk//bin [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error :  (in-process) [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error :  "C:/helix/work/correlation/build/wasi-sdk/bin/clang.exe" -cc1 -triple wasm32-unknown-wasi -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name aot-instances.dll.bc -mrelocation-model static -mframe-pointer=none -ffp-contract=on -fno-rounding-math -mconstructor-aliases -target-cpu generic -fvisibility=hidden -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=constructor -dwarf-version=4 -debugger-tuning=gdb -v -fcoverage-compilation-dir=C:/helix/work/workitem/e/publish -resource-dir C:/helix/work/correlation/build/wasi-sdk/lib/clang/16 -Oz -fdebug-compilation-dir=C:/helix/work/workitem/e/publish -ferror-limit 19 -fgnuc-version=4.2.1 -vectorize-slp -o "C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\tmpa2qi3t.tmp" -x ir "C:\\helix\\work\\workitem\\e\\wasm_build\\obj\\wasm\\for-build\\aot-instances.dll.bc" [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : clang -cc1 version 16.0.0 based upon LLVM 16.0.0 default target wasm32-wasi [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : LLVM ERROR: out of memory [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : Allocation failed [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script. [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : Stack dump: [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : 0.	Program arguments: C:\\helix\\work\\correlation\\build\\\\wasi-sdk\\\\bin\\clang.exe @C:\\helix\\work\\correlation\\build\\microsoft.netcore.app.runtime.wasi-wasm\\runtimes\\wasi-wasm\\native\\src\\wasi-default.rsp @C:\\helix\\work\\workitem\\e\\wasm_build\\obj\\wasm\\for-build\\wasi-compile-bc.rsp -c -o C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\tmpa2qi3t.tmp C:\\helix\\work\\workitem\\e\\wasm_build\\obj\\wasm\\for-build\\aot-instances.dll.bc [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : 1.	Code generation [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : 2.	Running pass 'Function Pass Manager' on module 'C:\helix\work\workitem\e\wasm_build\obj\wasm\for-build\aot-instances.dll.bc'. [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
C:\helix\work\correlation\build\wasm-shared\WasmApp.Common.targets(818,5): error : 3.	Running pass 'WebAssembly Instruction Selection' on function '@aot_instances_System_SZGenericArrayEnumerator_1_T_SINGLE__cctor' [C:\helix\work\workitem\e\publish\ProxyProjectForAOTOnHelix.proj]
```

**Comment by vargaz** [comment]
Those failures are happening on other PRs as well, so i think this is ok to merge.

**Pull request merged**